event.go

```diff
@@ -0,0 +1,225 @@
+		event.PrevEvents = emptyEventReferenceList
+	}
+	if event.AuthEvents == nil {
+		event.AuthEvents = emptyEventReferenceList
+	}
+	event.RawContent = rawJSON(event.content)
+	event.RawUnsigned = rawJSON(event.unsigned)
+	event.OriginServerTS = now.UnixNano() / 1000000
+	event.Origin = origin
+
+	// TODO: Check size limits.
+
+	var eventJSON []byte
+
+	if eventJSON, err = json.Marshal(&event); err != nil {
+		return
+	}
+
+	if eventJSON, err = addContentHashesToEvent(eventJSON); err != nil {
+		return
+	}
+
+	if eventJSON, err = signEvent(origin, keyID, privateKey, eventJSON); err != nil {
+		return
+	}
+
+	if eventJSON, err = CanonicalJSON(eventJSON); err != nil {
+		return
+	}
+
+	result.eventJSON = eventJSON
+	return
+}
+
+// NewEventFromUntrustedJSON loads a new event from some JSON that may be invalid.
+// This checks that the event is valid JSON.
+// It also checks the content hashes to ensure the event has not been tampered with.
+// This should be used when receiving new events from remote servers.
+func NewEventFromUntrustedJSON(eventJSON []byte) (result Event, err error) {
+	var event map[string]rawJSON
+	if err = json.Unmarshal(eventJSON, &event); err != nil {
+		return
+	}
+	// Synapse removes these keys from events in case a server accidentally added them.
+	// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/crypto/event_signing.py#L57-L62
+	delete(event, "outlier")
+	delete(event, "destinations")
+	delete(event, "age_ts")
+
+	// TODO: Check that the event fields are correctly defined.
+	// TODO: Check size limits.
+
+	if eventJSON, err = json.Marshal(event); err != nil {
+		return
+	}
+
+	if err = checkEventContentHash(eventJSON); err != nil {
+		result.redacted = true
+		// If the content hash doesn't match then we have to discard all non-essential fields
+		// because they've been tampered with.
+		if eventJSON, err = redactEvent(eventJSON); err != nil {
+			return
+		}
+	}
+
+	if eventJSON, err = CanonicalJSON(eventJSON); err != nil {
+		return
+	}
+
+	result.eventJSON = eventJSON
+	return
+}
+
+// NewEventFromTrustedJSON loads a new event from some JSON that must be valid.
+// This will be more efficient than NewEventFromUntrustedJSON since it can skip cryptographic checks.
+// This can be used when loading matrix events from a local database.
+func NewEventFromTrustedJSON(eventJSON []byte, redacted bool) Event {
+	return Event{redacted, eventJSON}
+}
+
+// Redacted returns whether the event is redacted.
+func (e Event) Redacted() bool { return e.redacted }
+
+// JSON returns the JSON bytes for the event.
+func (e Event) JSON() []byte { return e.eventJSON }
+
+// Redact returns a redacted copy of the event.
+func (e Event) Redact() Event {
+	if e.redacted {
+		return e
+	}
+	eventJSON, err := redactEvent(e.eventJSON)
+	if err != nil {
+		// This is unreachable for events created with EventBuilder.Build or NewEventFromUntrustedJSON
+		panic(fmt.Errorf("gomatrixserverlib: invalid event %v", err))
```

Comment:
I would make this even longer and splat the `e.eventJSON` that was passed in so we have more context other than "ruh roh this isn't JSON" or whatever comes out of `json.Unmarshal`.